### DISCUSSION TITLE
Mass matrix adaptation for chees-HMC

### DISCUSSION THREAD

**Discussion by jonny-so** [blackjax-devs/blackjax#757] [Q&A]
Hi,

I see in the implementation of chees-HMC adaptation the mass matrix is fixed to be the identity. I guess this is because the algorithm presented in the paper assumes the problem has been transformed such that the identity mass matrix is appropriate, but this was really just assumed for ease of exposition (see footnote 1 in the [chees-HMC paper](https://proceedings.mlr.press/v130/hoffman21a/hoffman21a.pdf)).  In practice it will be necessary adapt the mass matrix most of the time, alongside the step-size and trajectory length adaptation.

As a first attempt it would seem reasonable to use windowed adaptation to learn a mass matrix and (initial) step size, followed by a chees-HMC adaptation to learn the trajectory length function (and possibly further refine the step size).  But as far as I can tell this is not possible at the moment -- save for manually applying a linear change of variables to the problem -- as we can't pass a mass matrix in to the chees-HMC adaptation.  This is also suboptimal, as we would like to make use of a recent estimate of the optimal trajectory length while learning the mass matrix, and vice-versa - it seems as though in an ideal world we would incorporate trajectory-length adaptation into the windowed adaptation somehow.

Does anyone have any thoughts on this, or examples of using mass-matrix adaptation with chees-HMC more generally?

Thanks!

Jonny

**Comment by junpenglao** [comment]
+1 that we should extend the windowed adaptation to have more general mass matrix adaptation.

**Comment by junpenglao** [comment]
🤖 *This reply was drafted with Claude and reviewed by me (@junpenglao).*

Confirmed limitation: ChEES adaptation currently fixes the mass matrix to the identity (it assumes a pre-whitened target, per footnote 1 of the paper). For dense/low-rank mass-matrix adaptation today you can use `blackjax.window_adaptation` or `blackjax.window_adaptation_low_rank`. Generalizing ChEES warmup to co-adapt the mass matrix is still open — PR welcome.
